.github/workflows/check-code-quality.yml

```diff
@@ -31,3 +31,32 @@ jobs:
       uses: ludeeus/action-shellcheck@master
       with:
         severity: style
+
+    - name: Check for broken links
+      run: | 
+        touch valid_urls.txt redirected_urls.txt invalid_urls.txt
+        find . -type f \( -name "*.txt" -o -name "*.md" -o -name "*.py" -o -name "*.rst" \) | 
+          xargs grep -Eo "\b(https?|ftp):\/\/[-A-Za-z0-9+&@#\/%?=~_|$!:,.;]*[A-Za-z0-9+&@#\/%=~_|$]" | 
+          grep -v 'drive.google.com' |    # Returns 401 (needs google auth)
+          grep -v 'pipeline-hemis' |      # Returns 403 (private GH repo)
+          grep -v 'sciencedirect.com' |   # Returns 403
+          grep -v 'wiley.com' |           # Returns 403
+          grep -v 'sagepub.com' |         # Returns 403
+          grep -v 'ncbi.nlm.nih.gov' |    # Returns 403          
+          grep -v 'oxfordjournals.org' |  # Returns 403
+          grep -v 'docker.com' |          # Returns 403
+          grep -v 'ieeexplore.ieee.org' | # Returns 418
+          grep -v 'CHANGES.md' |          # Has too many GitHub links, triggers "Too many requests"
+          grep -v '%s' |                  # conf.py extlinks URL
+          sed 's/:/;/' | 
+          xargs -n 1 -P 8 ./.github/workflows/check-url.sh
```

Comment:
Is the `-P 8` really important for saving time? I ask because the manual page for `xargs -P` warns against running processes in parallel that write to `stdout` and/or to the same file, because the output can be garbled in that case.

But I can also imagine that some of the links are slow to fetch or time out, in which case it may be important

Reply:
Hmm... without `-P 8`, the workflow takes [about 13 minutes](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8194952477), compared to [~2m with -P 8](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8193300941). 

On the one hand, it's still less than our slowest tests. Plus, the linting checks will fail first (meaning that we'll still get pretty rapid feedback for everything but the broken links). 

I don't know if I feel strongly either way? Hmmm... What do you prefer between sensible stdout output vs. ~6x faster runtime?

Reply:
I think I might be leaning towards keeping `-P 8`? I've written the test such that the results are [summarized in a separate step](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8206297070/job/22445069619?pr=4395#step:8:17), anyway. Meaning that most of the time, we won't actually be looking at the in-process stdout of the `-P 8` step?

Reply:
I guess my main concern isn't `stdout`, but the three files `redirected_urls.txt`, `invalid_urls.txt`, `valid_urls.txt`, which get written to concurrently without synchronization. But looking at it again, even if they get garbled, I don't think it will affect the pass/fail nature of the check, so I guess it doesn't matter too much. The manual entry for `find -P` suggests having each concurrent process write to its own file, then merging them, but maybe that's too heavyweight for this little check.

Reply:
> I guess my main concern isn't `stdout`, but the three files `redirected_urls.txt`, `invalid_urls.txt`, `valid_urls.txt`, which get written to concurrently without synchronization.

Ahhh, I understand now! Thank you for clarifying, and my apologies for the misunderstanding. I hadn't thought about that, but you're absolutely right.

I guess if we wanted to protect the output while also running things in parallel, we would need to look at something like GNU parallel's [`sem`](https://stackoverflow.com/questions/42075841/bash-append-file-from-multiple-thread), then?

> But looking at it again, even if they get garbled, I don't think it will affect the pass/fail nature of the check, so I guess it doesn't matter too much.

Yes, that's true. The check will only fail if `invalid_urls.txt` has any contents at all.

In fact, these text files may be a red herring? I think I mainly used them for local debugging purposes, but they don't really affect how the CI check runs, as you noticed.